opentelemetry-sdk/src/log/log_processor.rs

```diff
@@ -0,0 +1,368 @@
+    shutdown: crossbeam_channel::Receiver<()>,
+}
+
+impl SimpleLogProcessor {
+    pub(crate) fn new(mut exporter: Box<dyn LogExporter>) -> Self {
+        let (log_tx, log_rx) = crossbeam_channel::unbounded();
+        let (shutdown_tx, shutdown_rx) = crossbeam_channel::bounded(0);
+
+        let _ = thread::Builder::new()
+            .name("opentelemetry-exporter".to_string())
+            .spawn(move || {
+                while let Ok(Some(log)) = log_rx.recv() {
+                    if let Err(err) = futures_executor::block_on(exporter.export(vec![log])) {
+                        global::handle_error(err);
+                    }
+                }
+
+                exporter.shutdown();
+
+                if let Err(err) = shutdown_tx.send(()) {
+                    global::handle_error(LogError::from(format!(
+                        "could not send shutdown: {:?}",
+                        err
+                    )));
+                }
+            });
+
+        SimpleLogProcessor {
+            sender: log_tx,
+            shutdown: shutdown_rx,
+        }
+    }
+}
+
+impl LogProcessor for SimpleLogProcessor {
+    fn emit(&self, data: LogData) {
+        if let Err(err) = self.sender.send(Some(data)) {
+            global::handle_error(LogError::from(format!("error processing log {:?}", err)));
+        }
+    }
+
+    fn force_flush(&self) -> LogResult<()> {
+        // Ignored since all logs in Simple Processor will be exported as they ended.
+        Ok(())
+    }
+
+    fn shutdown(&mut self) -> LogResult<()> {
+        if self.sender.send(None).is_ok() {
+            if let Err(err) = self.shutdown.recv() {
+                global::handle_error(LogError::from(format!(
+                    "error shutting down log processor: {:?}",
+                    err
+                )))
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// A [`LogProcessor`] that asynchronously buffers log records and reports
+/// them at a preconfigured interval.
+pub struct BatchLogProcessor<R: LogRuntime> {
+    message_sender: R::Sender,
+}
+
+impl<R: LogRuntime> Debug for BatchLogProcessor<R> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("BatchLogProcessor")
+            .field("message_sender", &self.message_sender)
+            .finish()
+    }
+}
+
+impl<R: LogRuntime> LogProcessor for BatchLogProcessor<R> {
+    fn emit(&self, data: LogData) {
+        let result = self.message_sender.try_send(BatchMessage::ExportLog(data));
+
+        if let Err(err) = result {
+            global::handle_error(err);
+        }
+    }
+
+    fn force_flush(&self) -> LogResult<()> {
+        let (res_sender, res_receiver) = oneshot::channel();
+        self.message_sender
+            .try_send(BatchMessage::Flush(Some(res_sender)))?;
+
+        futures_executor::block_on(res_receiver)
+            .map_err(|err| LogError::Other(err.into()))
+            .and_then(|identity| identity)
```

Comment:
Curious why we name it `identity`

Reply:
I believe it's a reference to the [identity function](https://en.wikipedia.org/wiki/Identity_function), even though the parameter is not the mapping in itself. I refactored this to just use `std::convert::identity` instead.